clippy_lints/src/indexing_slicing.rs

```diff
@@ -161,34 +187,34 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicing {
     }
 }
 
-/// Returns an option containing a tuple with the start and end (exclusive) of
-/// the range.
+/// Returns a tuple of options with the start and end (exclusive) values of
+/// the range. If the start or end is not constant, None is returned.
 fn to_const_range<'a, 'tcx>(
     cx: &LateContext<'a, 'tcx>,
     range: Range<'_>,
     array_size: u128,
-) -> Option<(u128, u128)> {
+) -> (Option<u128>, Option<u128>) {
```

Comment:
you could additionally return the span to be able to report the error directly on the offending value

Reply:
With the way the `to_const_range` is currently implemented, it returns `Some(0)` for the start value for things like `&x[..10]`, and it returns `Some(length_of_array)` for the end value for things like `&x[0..]`. 

Those `Some` values are then used to return early from the lint to avoid warning with `Consider using `.get(..n)`or `.get_mut(..n)` instead`. But in those cases, I don't think there is a span to return from `to_const_range`.

I suppose `to_const_range` could return an `Option<Span>`, but instead I just replaced the code which triggered the lint with this code (and something similar for `range.end` in the case where that is the out of bounds value):

```rust
utils::span_lint(
    cx,
    OUT_OF_BOUNDS_INDEXING,
    range.start.map_or(expr.span, |start| start.span),
    "range is out of bounds",
);
```

The trouble with this is that if both the start and end values are out of bounds, then only the start is reported by the lint. Of course, we could check for that condition and then fall back to simply reporting the lint on `expr.span`, but that starts to add significant complexity for a potentially small gain.

Let me know your thoughts on this. I'd be happy to add a commit to clarify what I am saying here if that would help, or to just move forward with the implementation if you think reporting directly on the offending value is worth the additional complexity in the code.

Reply:
You don't need an option, you can directly use `expr.span` inside this function instead of `None` when a `0` is magically created as the range start

Reply:
I just pushed a commit which directly reports the out of bounds value. 

Rather than adding a parameter to `to_const_range` for the `expr` (to use as a fallback value for start/end), and complicating the return tuple further, it seemed simpler to use `range.start.map_or(expr.span, |start| start.span)` when the lint is reported. Let me know if you disagree.